student/src/main/java/tds/student/web/handlers/TestShellHandler.java

```diff
@@ -231,31 +236,7 @@ private String parseOutLatencies (Map<String, String> formParams) {
 
     TestOpportunity testOpp = StudentContext.getTestOpportunity ();
 
-    // get responses
-    TestManager testManager = new TestManager (testOpp);
-    testManager.LoadResponses (true);
-
-    // check if test length is met
-    testManager.CheckIfTestComplete ();
-
-    if (!testManager.IsTestLengthMet ()) {
-      // TODO mpatel - Check to see status and message in response and make sure
-      // following works
-      String message = "Review Test: Cannot end test because test length is not met.";
-      _tdsLogger.applicationError (message, "reviewTest", request, null);
-      HttpContext.getCurrentContext ().getResponse ().sendError (HttpStatus.SC_FORBIDDEN, "Cannot end the test.");
-      return null;
-    }
-
-    // check if all visible pages are completed
-    PageList pageList = testManager.GetVisiblePages ();
-
-    if (!pageList.isAllCompleted ()) {
-      String message = "Review Test: Cannot end test because all the groups have not been answered.";
-      _tdsLogger.applicationError (message, "reviewTest", request, null);
-      HttpContext.getCurrentContext ().getResponse ().sendError (HttpStatus.SC_FORBIDDEN, "Cannot end the test.");
-      return null;
-    }
+    ResponseData<String> responseData = reviewTestService.reviewTest(testOpp, request);
```

Comment:
The validation I'm referring to in particular is lines 251 - 258 (!pageList.isAllCompleted), where the page groups are iterated over and checked for completion. I would imagine this is in place to prevent the student from moving to the review page without completing the last set of items. Thinking about it a bit further though, this may be slightly redundant. I recall you mentioning that segments are set to "satisfied" when items are being generated in the /updateResponse flow, but that wouldn't necessarily assure that the student actually completed/responded to all items.